lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1229 @@
+		auto found=this->ranges.find(keyToInsert);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second=ConstraintsLattice::updateRange(found->second, min, max);
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(
+				std::make_pair(keyToInsert,
+					createRange(
+						min.value_or(MIN),
+						max.value_or(MAX))));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value in the to_check structure
+	void ConstraintsLattice::insertInTocheck(const mlir::Value& value){
+		this->to_check.insert(value);
+	}
+
+	// This function is actually quite simpler since we always know 
+	// the lhs is not a constant and the rhs is a constant
+	// Result is useful to understand which check do to the operation
+	bool ConstraintsLattice::updateRelationalRange(mlir::Operation* statement, bool result, const mlir::Value& to_update, const mlir::Operation* other, const bool branch_to_update){
+
+		auto other_casted=mlir::dyn_cast<mlir::rlc::Constant>(other);
+
+		auto rhs_const=other_casted->getAttr("value").dyn_cast<IntegerAttr>().getInt();
+
+		std::optional<Integer> min;
+		std::optional<Integer> max;
+
+		llvm::TypeSwitch<mlir::Operation*>(statement)
+		.Case<mlir::rlc::GreaterOp>([&](mlir::rlc::GreaterOp op){
+			if(result) min=rhs_const+1;
+			else max=rhs_const;
+		})
+		.Case<mlir::rlc::LessOp>([&](mlir::rlc::LessOp op){
+			if(result) max=rhs_const-1;
+			else min=rhs_const;
+		})
+		.Case<mlir::rlc::GreaterEqualOp>([&](mlir::rlc::GreaterEqualOp op){
+			if(result) min=rhs_const;
+			else max=rhs_const-1;
+		})
+		.Case<mlir::rlc::LessEqualOp>([&](mlir::rlc::LessEqualOp op){
+			if(result) max=rhs_const;
+			else min=rhs_const+1;
+		});
+		
+		//llvm::outs()<<"should have changed\n";
+		// Insert the range where is needed
+		return this->insertOrUpdate(std::make_pair(to_update,branch_to_update),min,max);
+	}
+
+	//////////////////////////////////////////////////////////////////////////////
+	// ConstraintsAnalysis METHODS
+	//////////////////////////////////////////////////////////////////////////////
+
+	// Definitions taken from https://mlir.llvm.org/doxygen/DenseAnalysis_8h_source.html#l00423
+
+	void ConstraintsAnalysis::visitOperation(
+		mlir::Operation *op, 
+		const ConstraintsLattice &after,
+		ConstraintsLattice *before){
+
+		//llvm::outs()<<"visiting operation: ";
+		//op->print(llvm::outs());
+		//llvm::outs()<<"\n";
+
+		// For the moment implement the basic operations
+		
+		// We have to decide which return values to update (true or false)
+		// In order to do so we need to :
+		// - peek at the successors of this block 
+		// - look at their lattice
+		// - find in their lattice if there is some value with key true/false
+		// - keep in mind that information
+
+		llvm::TypeSwitch<mlir::Operation*>(op)
+		.Case<mlir::rlc::AddOp>([&](mlir::rlc::AddOp op){
```

Comment:
this switch should not be needed. the point of doing the interface is that you should be able to do
```
if (auto casted = op.dynamic_cast<mlir::rlc::ConstraintAnalyzable>()) {
  casted.analyzeConstraint(...):
}
else {
 // something conservative
}
```

Reply:
That is what I wanted to do too, to declare a super-interface (which is still there but commented).

But I think the only solution with that is to implement also the methods for `mlir::rlc::ConstraintsAnalyzable` which creates a bit of redundancy but i think its doable, it is what i wanted to look into precisely.

And also it would be better for me to have another layer of abstraction, because for terminator operations I would need to call the join function of the lattices, and for the nonterminator (arithmetic, builtin-assign) i would simply need to call a copy of the lattices